docs: rbac getting started guide

mmcallister opened the pull request:
Setting up a new PR from rbac guide issue #46417. 

Discuss core concepts of user and role #15156, provide simple rbac starter examples  #39241.
Revising existing content, adding an introduction and basics that were missing in the original and how-to guide for setting up RBAC, with additional examples for kubernetes

mmcallister committed:
revising rbac guide issue 46417

mmcallister committed:
link edits

mmcallister committed:
space edits

mmcallister commented:
@ptgott could you have a look at this again when you have time? thank you!

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
 
-We will start with local users and preset roles, assign roles to SSO users, and
-wrap up with creating your own role.
+## What is RBAC in Teleport?
```
```suggestion
```
The first couple of sections are very short, so I think we can consolidate them.

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
+## What is RBAC in Teleport?
 
-## Prerequisites
+In Teleport, any local, SSO, or robot user can be assigned one or several roles. Roles govern access to your resources and infrastructure.
```
```suggestion
A Teleport user can be assigned one or several roles. Roles govern access to your resources and infrastructure.
```
Since we don't introduce the distinction between kinds of users until later in this page

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
+Each role defines specific permissions to ensure that users only have access to what they need. 
 
-- (!docs/pages/includes/tctl.mdx!)
+## What is a role?
```
```suggestion
```
Since the first few sections are short, I think we can consolidate them.

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
+External users are useful for larger organizations or those that want to centralize identity management via an existing SSO provider. This is often preferred for enterprise environments where identity management is handled outside Teleport.
 
-</Tabs>
+### Robots
```
The Machine ID section of the docs calls these "bot users".

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
-</TabItem>
-<TabItem scope={["enterprise", "cloud"]} label="Commercial">
-Invite the local user Alice as cluster `editor` and `reviewer`:
+### Internal users
```
I would standardize on "local" and "SSO" users instead of "internal" and "external":

```
grep -hREoi "(internal|external|local|SSO) user" docs/pages | tr '[:upper:]' '[:lower:]'| sort | uniq -c
  13 external user
   1 internal user
 131 local user
  29 sso user
```

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
-</Tabs>
+Some examples of roles include:
 
-Because Alice has two or more roles, permissions from those roles create a union. She
-will be able to act as a system administrator and auditor at the same time.
+- **access**: Grants basic access to resources.
+- **editor**: Allows users to modify configuration settings.
+- **auditor**: Allows users to view logs and auditing activities.
```
Is there a way to merge this with our table of preset roles? Otherwise, it's not clear to me what we're referring to when we talk about these particular roles

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
 
-Next, follow the instructions to set up an authentication connector that maps
-users within your SSO solution to Teleport roles.
+## Understanding Teleport’s Resource Tagging and Matching
```
We use sentence case in headings: https://github.com/gravitational/docs/blob/main/docs-contributors/style-guide.md#headings

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
-Create a SAML or OIDC application that Teleport can integrate with, then
-create an authentication connector that maps users within your application to
-Teleport roles.
+- Resource Tagging: Teleport uses labels (key-value pairs) to tag resources like SSH nodes, databases, or Kubernetes clusters. For example, a server might have the label env: production.
```
```suggestion
- Resource Tagging: Teleport uses labels (key-value pairs) to tag resources like SSH nodes, databases, or Kubernetes clusters. For example, a server might have the label `env: production`.
```
Otherwise, this is difficult to read as regular text

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
-create an authentication connector that maps users within your application to
-Teleport roles.
+- Resource Tagging: Teleport uses labels (key-value pairs) to tag resources like SSH nodes, databases, or Kubernetes clusters. For example, a server might have the label env: production.
+- Role Matching: A role in Teleport allows or denies access to resources based on matching these labels. For instance, if a user has a role that allows access to resources labeled env: staging, they will only access resources with that label.
```
```suggestion
- Role matching: A role in Teleport allows or denies access to resources based on matching these labels. For instance, if a user has a role that allows access to resources labeled `env: staging`, they will only access resources with that label.
```
- Using code style for key/value pairs
- "Role matching" isn't a proper noun

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
+By creating and assigning roles, administrators can enforce security policies, manage access, and ensure users have the appropriate level of privileges based on their responsibilities.
 
-Teleport provides several preset roles:
+## Teleport users
```
Would it make sense to focus this guide on roles and label matching and, rather than spell out the kinds of users here, link to the [User Types](https://pr-48029.d3pp5qlev8mo18.amplifyapp.com/docs/reference/user-types) reference in a "Further reading" or "Next steps" section?

ptgott replied:
@mmcallister What do you think about the comment above? In general, I think this guide should be a lot more focused. I would start by identifying the objective of the guide in the intro paragraph (making it as narrow as possible), then editing the structure of the guide based on that objective. How does that sound?

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
 
-Save the file below as `okta.yaml` and update the `acs` field.
-Any member in Okta group `okta-admin` will assume a built-in role `admin`.
+Teleport’s roles are defined using YAML. Here’s a breakdown of a basic role configuration:
```
`tctl` reads YAML, but you can also manage roles with Terraform, the Teleport Kubernetes operator, and other API clients.

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
+- **allow**: Lists the permissions the role grants, such as allowed logins or access to resources based on labels.
+- **deny**: Specifies any permissions that should be explicitly denied, even if they would otherwise be allowed.
+
+##  Creating a Basic Role for SSH Access
```
I would make the guide more focused as a how-to guide or a conceptual guide, rather than bringing together features of both of them (see the [style guide](https://github.com/gravitational/docs/blob/main/docs-contributors/style-guide.md#organizing-information-into-purpose-driven-documents)).

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
-Create the `oidc` resource:
+Alice can now log in to staging servers, but she will be blocked from production environments.
+
+## Add local users with preset roles
```
I'm not sure about the place of this section within the guide. Is it meant as a set of instructions or a conceptual description?

It's possible to map an SSO user to preset roles, so if we want to use this section to lay out the preset roles, I would make it about that and remove the local user aspect.

ptgott commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,193 +1,172 @@
-## Step 3/3. Create a custom role
+Because Alice has two or more roles, permissions from those roles create a union. She will be able to act as a system administrator and auditor at the same time.
+
+## Create a custom role
```
Same question as above re: the "Add local users with preset roles" section. I am not sure whether to treat this as a conceptual description or a how-to instruction. Is this an example of using label matching as the guide describes earlier?

mmcallister replied:
The idea was to provide basic steps to get started and to get new users familiar with these topics, such as labeling, and access.

ptgott commented on docs/pages/admin-guides/access-controls/resource-rbac.mdx:
```diff
@@ -0,0 +1,89 @@
+description: Implementing fine-grained access control with RBAC.
+---
+
+Per-resource RBAC enables administrators to control user access on a per-resource basis. Using a granular approach allows you to grant more precise management of user permissions in your environment.
+
+## What is per-resource RBAC?
+
+This capability is particularly useful in complex, multi-tenant environments where user groups have varying levels of access requirements. It allows for precise control over permissions, enabling restricted access to sensitive resources for certain users while granting broader access to others based on their operational needs.
```
I would lay out a more explicit definition of per-resource RBAC. Right now, the guide only talks about what it allows, not what it is. I also think we can remove the first heading, since both of these paragraphs can work as introductory text for the guide.

mmcallister replied:
Removed initial heading and added a definition

ptgott commented on docs/pages/admin-guides/access-controls/resource-rbac.mdx:
```diff
@@ -0,0 +1,89 @@
+
+This capability is particularly useful in complex, multi-tenant environments where user groups have varying levels of access requirements. It allows for precise control over permissions, enabling restricted access to sensitive resources for certain users while granting broader access to others based on their operational needs.
+
+## Use cases
```
So far, this guide describes the benefits of per-resource RBAC extensively, including in this use cases section, but we haven't gotten into what per-resource RBAC is yet. 

I think once the reader has a more concrete idea of what per-resourcee RBAC is, we can spend less time talking about potential use cases. I.e., if a reader identifies their own use case, they don't need to read about use cases that don't apply to them.

ptgott commented on docs/pages/admin-guides/access-controls/resource-rbac.mdx:
```diff
@@ -0,0 +1,89 @@
+In CI/CD pipelines, different stages may require different levels of access; per-resource RBAC can be used to grant just-in-time access for specific tasks in the pipeline.
+
+Let's look at some examples of how per-resource RBAC can be implemented:
+
```
Would it make sense to merge this information into the role reference or the RBAC getting started guide? Aside from the description of benefits and use cases, it seems like the examples below could work in one of those guides.

mmcallister replied:
I didn't want to make the getting started page too long, but the content does work well together. Removed some of the redundancies and combined them, lmk what you think

mmcallister committed:
Update docs/pages/admin-guides/access-controls/getting-started.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

mmcallister committed:
Update docs/pages/admin-guides/access-controls/getting-started.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

mmcallister committed:
Update docs/pages/admin-guides/access-controls/getting-started.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

mmcallister committed:
rbac intro - post-feedback

mmcallister committed:
vale prose niggles

mmcallister commented:
[Latest preview link](https://docs-onobql29z-goteleport.vercel.app/docs/admin-guides/access-controls/getting-started/)

ptgott commented:
Hi @mmcallister, what do you think of the comments I left in [this thread](https://github.com/gravitational/teleport/pull/48029/files#r1823255607)?

mmcallister committed:
post-feedback - overhaul

ptgott committed:
Edit the Access Controls getting started guide (#48759)

* Edit the Access Controls getting started guide

1. Edit the introduction.

   - Remove unnecessary sentences
   - Move the "What is RBAC" text to the introduction so we describe RBAC
     before saying that this guide shows you how to use it.

2. Move the outline of RBAC components into a "How it works" section.
   This follows the structure of other how-to guides and adds some
   context by using prose.

3. Change the step-by-step instructions.

   - Make these first in the guide to keep with the conventions for our
     step-by-step guides.
   - Add steps to enroll two servers as Docker containers in order to
     demonstrate RBAC.

4. Add conceptual sections under the step-by-step instructions in a
   "Next steps" section. This follows the convention for our how-to
   guides.

   - Reformat "Maintaining your labeling strategy" to make it more
     compact.
   - Edit style in the use cases section to use headings and add a short
     introduction to each code block.

5. Remove the per-resource RBAC section. This section combines Database
   Access Controls and Kubernetes per-resource RBAC. We use
   "per-resource RBAC" to refer to Kubernetes resource RBAC. Instead,
   add a section that links to RBAC-related guides for each kind of
   resource.

* Apply suggestions from code review

Co-authored-by: marie <marie.mcallister@goteleport.com>

---------

Co-authored-by: marie <marie.mcallister@goteleport.com>

ptgott commented on docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx:
```diff
@@ -0,0 +1,138 @@
+In Teleport's flexible RBAC system, you can define custom roles and granularly control what users can do within your Kubernetes clusters. 
+Limit access to specific namespaces, enforce read-only permissions, or restrict administrative actions as needed.
+
+:Note:
+Please reference [Setting Up Teleport Access Controls for Kubernetes](../../enroll-resources/kubernetes-access/manage-access.mdx) before proceeding, if you have not done so already.
```
```suggestion
<Notice type="tip">
Reference [Setting Up Teleport Access Controls for Kubernetes](../../enroll-resources/kubernetes-access/manage-access.mdx) before proceeding, if you have not done so already.
</Notice>
```
- Using our docs engine syntax to create a notice
- Removing "Please", which I don't think we need in technical writing

ptgott commented on docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx:
````diff
@@ -0,0 +1,138 @@
+
+Below is an example of a role that provides developers with access to only the dev namespace:
+
+```code
````
```suggestion
```yaml
```

ptgott replied:
We should use `yaml` for other YAML examples as well so syntax highlighting works.

ptgott reviewed (approved):
Approved with suggestions

ptgott commented:
@mmcallister Just checking to see if we're still planning to move ahead with this one. Let me know if you need any help shepherding it through!

ptgott commented:
> @mmcallister Just checking to see if we're still planning to move ahead with this one. Let me know if you need any help shepherding it through!

@mmcallister Following up on this. Thanks!

mmcallister committed:
Update docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

mmcallister committed:
Update docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
+directly to users, you create roles that define sets of permissions, then assign
+those roles to users.
+
+This guide helps you understand and implement Role-Based Access Control (RBAC)
```
I think the most common thing is to spell out the acronym on first use, and then use the acronym going forward.

I would update line 7 tog spell out RBAC before using the acronym, and then use only the acronym here.

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
+       method: token
+     proxy_server: <Var name="example.teleport.sh:443" />
+   auth_service:
+     enabled: "no"
```
I think we just merged a PR to change these yes/no values to true/false. Let's not undo those changes by adding more of the old style.

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
````diff
@@ -1,253 +1,396 @@
+   enrolled:
+
+   ```code
+   $ tsh ls env=local-dev
````
Do we need to tell/teach the user about `tsh login` first, or do we assume they've already logged in?

mmcallister replied:
Added a bit for this

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
````diff
@@ -1,253 +1,396 @@
+   $ tctl tokens add --type=node
+   ```
+
+1. Assign the token to <Var name="token2" />.
````
Would it simplify the guide to reuse the same token?

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
+   kind: role
+   version: v7
+   metadata:
+     name: local-dev-access
```
Maybe we should call this `local-dev-only`, because the deny rule means the user will never be able to access prod.

Or maybe we should change the label matcher to `env: local-dev` and remove the deny rule to make the first role the user develops as simple as possible.

mmcallister replied:
Removed the deny rule

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
+     name: local-dev-access
+   spec:
+     allow:
+       logins: ['root']
+       node_labels:
+         'env': 'local-*'
+     deny:
+       node_labels:
+         'env': 'local-prod'
```
```suggestion
       logins: [root]
       node_labels:
         env: local-*
     deny:
       node_labels:
         env: local-prod
```

I don't think any of these quotes are necessary.

mmcallister replied:
Deleted quotes

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
+version: v7
 metadata:
-  name: okta
+  name: senior-developer
```
```suggestion
  name: backend-developer
```

In most organizations, access is based on what team you are part of, not your job level.

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
-(!docs/pages/includes/create-role-using-web.mdx!)
-
-## Next steps
+This role allows a user to read audit log entries:
```
I don't think this is correct. The `session` resource is related to active sessions, but it's not used much in modern Teleport. I'd probably just eliminate this example to avoid opening up that can of worms.

Plus, Teleport comes with a preset `auditor` role so it's a bit awkward to provide an example that conflicts with the existing name.

mmcallister replied:
Deleted this example

zmb3 commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,253 +1,396 @@
+
+### Best practices
+
+Keep these key principles in mind to maintain a secure environment:
```
This section seems duplicative with lines 316-320.

mmcallister replied:
There's a little overlap but the preceding lines were specifically about labeling strategy vs general guidelines. If you think this isn't useful though I will delete it.

zmb3 commented on docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx:
```diff
@@ -0,0 +1,139 @@
+Limit access to specific namespaces, enforce read-only permissions, or restrict administrative actions as needed.
+
+<Admonition type="tip">
+Reference [Setting Up Teleport Access Controls for Kubernetes](../../enroll-resources/kubernetes-access/manage-access.mdx) before proceeding, if you have not done so already.
```
```suggestion
Reference [Setting Up Teleport Access Controls for Kubernetes](./manage-access.mdx) before proceeding, if you have not done so already.
```

Wouldn't this be identical?

mmcallister committed:
Update docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

mmcallister committed:
post-Zac-feedback

mmcallister commented:
Hi @ptgott Can we revisit this when you have the bandwidth? All feedback was addressed, and I believe this information would still be helpful. Ty!

ptgott committed:
Minor edits

- Introduce the example at the top of the Prerequisites section.
- Fix a link path.
- Use more consistent parts of speech in H3s within "Next steps".

ptgott reviewed (commented):
Gave this a quick look and fixed some minor issues with [fb26073](https://github.com/gravitational/teleport/pull/48029/commits/fb26073a4acf4e715bb5e8a0f51d9fdf283118dc). Also left a question.

I think my approval is still in the books here—we'll just need a G1 approval as well!

nklaassen commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
````diff
@@ -1,273 +1,378 @@
 
-## Step 2/3. Map SSO users to roles
+   ```code
+   $ tsh ssh root@<Var name="ba2290caf694" />
````
do our docker containers actually allow SSH access? i get this error
```
$ tsh ssh root@8db48991d7b6
Failed to launch: fork/exec /sbin/nologin: no such file or directory.
ERROR: Process exited with status 255
```

nklaassen commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,273 +1,378 @@
-<Tabs>
-<TabItem scope={["oss"]} label="Teleport Community Edition">
-Invite the local user Alice as cluster `editor`:
+   However, since the `deny.node_labels` field specifies `env:local-prod`, a
+   user with this role would only be able to access the server with the
+   `env:local-dev` label.
```
i'm not suggesting adding it, but the example role doesn't actually contain any `deny` section, we could just say they would only be allowed to access nodes with `env:local-dev`

nklaassen commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
````diff
@@ -1,273 +1,378 @@
+   ```
 
-You can update the user's roles using the `tctl users update` command:
+   Since `alice` is denied access to servers with the `env:local-prod` label,
````
```suggestion
   Since `alice` is not allowed access to servers with the `env:local-prod` label,
```

nklaassen commented on docs/pages/admin-guides/access-controls/getting-started.mdx:
```diff
@@ -1,273 +1,378 @@
+## Next steps
 
-### Teleport Enterprise
+In this guide, you created a Teleport role that allowed and denied access to SSH
```
```suggestion
In this guide, you created a Teleport role that allowed access to SSH
```

nklaassen commented on docs/pages/enroll-resources/kubernetes-access/kubernetes-access-examples.mdx:
````diff
@@ -0,0 +1,139 @@
+
+In organizations with DevOps or platform engineering teams, certain users need full administrative access to all Kubernetes clusters for management purposes. This role grants full access to all Kubernetes resources across all clusters, allowing the user to perform any action on any resource.
+
+```code
+kind: role
+version: v6
+metadata:
+  name: admin-k8s-access
````
might want to update this to role v7 now https://goteleport.com/docs/enroll-resources/kubernetes-access/manage-access/#define-a-role

mmcallister commented:
This is outdated, and we have work in progress for the enroll-resources guides to improve agentic workflows. A lot of improvements have since been made to rbac/getting-started - https://goteleport.com/docs/zero-trust-access/rbac-get-started/
